### PR TITLE
Enable NHWC transformer when generating ORT format model

### PR DESCRIPTION
**Pull request by skottmckay**
**Description**: 
We want to apply the new NHWC transformer when creating a model to run using the CPU EP on an ARM device. 

We do not however want the NCHWc transformer which is also at the same optimization level to run when creating an ORT format model. The NCHWc transformer will use information from the device creating the model (which is unlikely to match the mobile device the model will be run on, making it invalid), and targets x64 not ARM. 

* Allow specific optimizers to be disabled.
  - replace unused ability to specify just the optimizers to run
* Allow the disabled list to be specified via the python bindings
  - expected usage is internal, so using kwargs so as not to pollute the documentation for creating an InferenceSession in python with stuff no user is likely to need
* Update the ORT format model conversion script to disable NCHWc transformer when level is 'all'
  - currently there aren't any known use cases where we'd want the NCHWc transformations to run as they create a device specific model and aren't used on ARM
    - the ORT format model is not expected to be generated on the target device (e.g. generate on Windows/Linux/macOS to deploy to Android/iOS so there's a good chance we'd generate a useless/invalid model
  - default to 'all' as both ARM and MLAS prefer NHWC, and that level is required for the NHWC transformer to run

**Motivation and Context**
Allows additional performance optimizations to be applied.

### CODE REVIEW

File: onnxruntime/core/session/inference_session.h [hunk 1]
```diff
@@ -178,12 +178,17 @@ class InferenceSession {
...
+    * For an optimizer to be enabled, it must be allowed at the current optimization level (as specified in
+    * session options), and NOT in optimizers_to_disable. 
+    * This allows finer grained control of the enabled/disabled optimizations.
+    * Must be called before Initialize() to take effect.
+    * 
     * Calling this API is optional.
-    * When this list is provided ORT ignores the levels set in session options.
     * @return OK if success.
     */
-  common::Status AddCustomTransformerList(const std::vector<std::string>& transformers_to_enable) ORT_MUST_USE_RESULT;
```
Comment (gwang-msft) [inline onnxruntime/core/session/inference_session.h:186]: >common::Status [](start = 2, length = 14)

change to void?

File: onnxruntime/core/session/inference_session.h [hunk 2]
```diff
@@ -178,12 +178,17 @@ class InferenceSession {
...
+  common::Status FilterEnabledOptimizers(const std::unordered_set<std::string>& optimizers_to_disable)
```
Comment (gwang-msft) [inline onnxruntime/core/session/inference_session.h:190]: >common::Status [](start = 2, length = 14)

change to void?

Comment (skottmckay) [reply onnxruntime/core/session/inference_session.h:190]: Considered that but as it's going to be exposed publicly I'd prefer the flexibility to return errors in the future.

File: orttraining/orttraining/test/optimizer/graph_transformer_utils_test.cc [hunk 3]
```diff
@@ -20,25 +20,36 @@ TEST(GraphTransformerUtilsTestsForTraining, TestGenerateGraphTransformers) {
...
+      auto l1_rewrite_rules = optimizer_utils::GenerateRewriteRules(TransformerLevel::Level1);
+      ASSERT_TRUE(rule_transformer->RulesCount() == l1_rewrite_rules.size() - 1);
+      break;
     }
   }
-  ASSERT_TRUE(rule_transformer && rule_transformer->RulesCount() == 1);
 
-  transformers = training::transformer_utils::GenerateTransformers(TransformerLevel::Level2, {}, {}, custom_list);
+  ASSERT_TRUE(rule_transformer) << "RuleBased transformer should have been added by GenerateTransformers";
+
```
Comment (gwang-msft) [inline orttraining/orttraining/test/optimizer/graph_transformer_utils_test.cc:47]: Move this after line 37? Actualy can this ever be null?

**Review by gwang-msft — APPROVED** [review]
:shipit:

File: orttraining/orttraining/test/optimizer/graph_transformer_utils_test.cc (hunk 3 above)
Comment (skottmckay) [reply orttraining/orttraining/test/optimizer/graph_transformer_utils_test.cc:47]: It's null if we didn't add a rule based transformer to the list in GenerateTransformers. The 'rule' we're matching (EliminateIdentity) is inside of a transformer called L1_RuleBasedTransformer.

**Comment by skottmckay** [comment]
@tianleiwu and @SherlockNoMad: I also applied the changes to some training code as that uses the same utils to generate the transformers, and inherits from InferenceSession.h. I didn't see any usage of the custom list (and don't really understand why you'd want to manually list all the optimizers to enable) in the training code base but just checking it's okay to make the change from 'only use optimizers on this list' to 'exclude optimizers on this list'. Combined with the changes in #7131 to allow manually enabling GeluApproximation I think this is a more useful setup.

**Review by gwang-msft — APPROVED** [review]

**Pull request merged**